### PR TITLE
Use EntityUpdateMixin, update entity fields

### LINKED ISSUE #36: Research Realm.realm_proxy
The semantics of the `Realm.realm_proxy` field are not well understood. As a result, the field is omitted. The API documentation mentions it, though. From the API documentation:

> POST /api/realms
> realm[realm_proxy_id]
> Proxy to use for this realm
>  Value: number.

Research the semantics of this field and add it to `nailgun.entities` if appropriate.

### LINKED ISSUE #52: Add `realm` field to Location entity class
The [Location](http://nailgun.readthedocs.org/en/latest/api/nailgun.html#nailgun.entities.Location) entity does not have a `realm` field. The API docs list this field, and the field is listed as a comment in the class definition itself. However, the field is currently commented out.

The reason is that, when reading a location, no realm information is returned by the server. See [BZ 1216234](https://bugzilla.redhat.com/show_bug.cgi?id=1216234). However, it may be possible to work around this issue. For example, method [`UserGroup.read`](http://nailgun.readthedocs.org/en/latest/api/nailgun.html#nailgun.entities.UserGroup.read) works around an _extremely_ similar issue. Perhaps the solution used in that method can be used here as well.

### LINKED ISSUE #56: Add `discovery` field to Subnet entity class
The [Subnet](http://nailgun.readthedocs.org/en/latest/api/nailgun.html#nailgun.entities.Subnet) entity does not have a `discovery` field. Even though the API docs don't list this field, using the web ui shows that one can **POST** this attribute but cannot **GET** it..

The reason is that, when reading a subnet, no discovery proxy information is returned by the server. See [BZ 1217146](https://bugzilla.redhat.com/show_bug.cgi?id=1217146).

### LINKED ISSUE #75: Research Subnet.discovery
The UI allows you to associate a subnet with a "Discovery Proxy". Here's a screenshot demonstrating the capability:

screenshot_2015-05-15_12-57-20

When you hit the "submit" button, a pseudo HTTP POST request is submitted to `https://<hostname>/subnet/<subnet-id>-<subnet-name>`. Here's a subset of the payload:

```json
{
    "_method": "put",
    "subnet": {"discovery_id": 1},
}
```

"discovery" appears to be a property of the Subnet entity. Can we add this proeprty to the entity? Unfortunately, no. There's a couple big problems:

* HTTP GET responses don't include a "discovery" or "discovery_id" attribute.
* HTTP PUT calls appear to ignore any "discovery_id" attribute that is submitted.
* HTTP PUT responses don't include a "discovery_id" attribute.

What to do? I'm not sure.

### PR DESCRIPTION
**Pull request by Ichimonji10**
Add the `EntityUpdateMixin` to several entities. Add methods to those entities
as needed in order to make the `update*` methods work correctly.

Add fields to entities:

* `ContentView.puppet_module`
* `ContentView.version`
* `ContentViewVersion.environment`
* `ContentViewVersion.puppet_module`
* `Location.realm` (Fix #52)
* `Organization.compute_resource`
* `Organization.config_template`
* `Organization.domain`
* `Organization.environment`
* `Organization.hostgroup`
* `Organization.media`
* `Organization.realm`
* `Organization.smart_proxy`
* `Organization.subnet`
* `Organization.subnet`
* `Organization.user`
* `Product.repository`
* `Realm.location`
* `Realm.organization`
* `Realm.realm_proxy` (Fix #36)
* `Subnet.discovery` (Fix #56, Fix #75)
* `Subscription.product`

Change how the `create_payload` method works. Assuming the server returns a JSON
payload of `{'one_to_one': None}`, the following now holds true:

    >>> class SomeEntity(…):
    ...     def __init__(…):
    ...         self._fields = {'one_to_one': OneToOneField(OtherEntity)}
    ...
    >>> SomeEntity(id=1).read().one_to_one
    None

Formerly, the `one_to_one` attribute would be an `OtherEntity` object with an ID
of `None`. The former behaviour is non-sensical.

Expand test coverage. Test coverage should now be at least 90%.

### CODE REVIEW

**Comment by Ichimonji10** [comment]
I used NailGun in this state when creating the test results documented in robottelo#2441. In other words, please look at robottelo#2441 for proof that this change request is beneficial and does not totally bork things.

**Comment by Ichimonji10** [comment]
I'm going to release NailGun 0.17.0 after this PR is merged. Notice that I have *not* added the new EntityUpdateMixin to all of the entity classes. I've only added the EntityUpdateMixin to the entities that robottelo updates. The EntityUpdateMixin should be added to all appropriate entities in 0.18.0.

**Comment by JacobCallahan** [comment]
ACK

File: nailgun/entities.py [hunk 1]
```diff
@@ -619,6 +638,15 @@ def read(self, entity=None, attrs=None, ignore=()):
             attrs['template_kind'] = {'id': template_kind_id}
         return super(ConfigTemplate, self).read(entity, attrs, ignore)
 
+    def update(self, fields=None):
+        """Fetch a complete set of attributes for this entity.
+
+        FIXME: File a bug at https://bugzilla.redhat.com/ and link to it.
```
Comment (oshtaier) [inline nailgun/entities.py]: Is it possible to fix that before merge?

File: nailgun/entities.py [hunk 2]
```diff
@@ -1964,8 +2027,16 @@ def create(self, create_missing=None):
...
         return Location(self._server_config, id=attrs['id']).read()
 
-    def read(self, entity=None, attrs=None, ignore=()):
-        """Compensate for the pluralization of several fields."""
+    def read(self, entity=None, attrs=None, ignore=('realm',)):
+        """Work around several bugs in the server's response.
+
+        * Compensate for the pluralization of the "smart_proxy" and "media"
+          fields.
+        * Do not try to read oany of the attributes listed in the ``ignore``
```
Comment (oshtaier) [inline nailgun/entities.py]: Typo: 'Do not try to read oany'

File: nailgun/entities.py [hunk 3]
```diff
@@ -2740,17 +2870,41 @@ def read(self, entity=None, attrs=None, ignore=()):
...
     def __init__(self, server_config=None, **kwargs):
         self._fields = {
+            'location': entity_fields.OneToManyField(Location),
             'name': entity_fields.StringField(required=True),
+            'organization': entity_fields.OneToManyField(Organization),
+            'realm_proxy': entity_fields.OneToOneField(
+                SmartProxy,
+                required=True,
+            ),
             'realm_type': entity_fields.StringField(required=True),
```
Comment (oshtaier) [inline nailgun/entities.py]: maybe we can do something like this:
```
'realm_type': entity_fields.StringField(
                choices=(
                    'Red Hat Identity Management',
                    'Active Directory',
                ),
                required=True,
            ),
```

File: nailgun/entities.py [hunk 4]
```diff
@@ -2740,17 +2870,41 @@ def read(self, entity=None, attrs=None, ignore=()):
         return super(PuppetModule, self).read(entity, attrs, ignore)
 
 
-class Realm(Entity):
+class Realm(
```
Comment (oshtaier) [inline nailgun/entities.py]: Was you able to create 'Realm' entity with code we have here?
When I tried I get: ```<422: "URL is invalid - only  http://, https:// are allowed">```

If I try to create SmartProxy entity I get:
```
"Unable to communicate with the proxy: ERF12-2530 [ProxyAPI::ProxyException]: Unable to detect features ([RestClient::RequestTimeout]: Request Timeout) for proxy http://test.org/features","Please check the proxy is configured and running on the host."]
```

Maybe, I need full set of your changes :). Appreciate your feedback on that

File: nailgun/entity_mixins.py [hunk 5]
```diff
@@ -299,11 +311,14 @@ def __init__(self, server_config=None, **kwargs):
         for field_name, field_value in kwargs.items():  # e.g. ('admin', True)
             field = self._fields[field_name]  # e.g. A BooleanField object
             if isinstance(field, OneToOneField):
-                setattr(self, field_name, _make_entity_from_id(
-                    field.gen_value(),
-                    field_value,
-                    self._server_config
-                ))
+                if field_value is None:
+                    setattr(self, field_name, field_value)
```
Comment (oshtaier) [inline nailgun/entity_mixins.py]: :+1:

**Comment by oshtaier** [comment]
Basically, it is amazing PR which improving nailgun a lot
ACK pending few questions

File: nailgun/entities.py (hunk 3 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: Thanks for this! I'll definitely make this fix.

File: nailgun/entities.py (hunk 2 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: Whoops! I'll fix this.

File: nailgun/entities.py (hunk 1 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: There are quite a few work-arounds like this. I forgot to file bugs for each of these FIXMEs, though. I'll file a bug asking for these FIXMEs to be taken care of.

File: nailgun/entities.py (hunk 4 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: I did manage to create and work with a 'Realm' entity via the API. Let me spend a few minutes working on this, and I'll see if I can post some output showing how I did it.

File: nailgun/entity_mixins.py (hunk 5 above)
Comment (Ichimonji10) [inline nailgun/entity_mixins.py]: :smile:

This definitely seems like more sane behaviour. I don't know what I was smoking when I hammered out the original design.

Actually, I *do* know what I was smoking. I was smoking the "I'm-not-unit-testing-everying" pipe. The process of expanding unit test coverage is what really forced me to redesign this. This is yet another argument in favor of developers writing tests as part of  their workflow.

File: nailgun/entities.py (hunk 4 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: I can create a Realm with the current code. There's a catch, unfortunately - the `create` method is broken. See the example code below.

The fragility of NailGun with regards to the names that the server may return is very annoying. I've described the issue in #110 and #114. I've already implemented a fix, too, in my "robustness" branch.

```python
>>> import robottelo; from nailgun import entities
>>> smart_proxy = entities.SmartProxy(id=1).read()
>>> realm = entities.Realm(
...     realm_proxy=smart_proxy,
...     realm_type='Active Directory',
...     name='foo',
... ).create(create_missing=False)
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 766, in create
    return self.read(attrs=self.create_json(create_missing))
  File "/home/ichimonji10/code/nailgun/nailgun/entities.py", line 2909, in read
    return super(Realm, self).read(entity, attrs, ignore)
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 621, in read
    in attrs[field_name + 's']  # e.g. "users"
KeyError: 'organizations'
>>> attrs = entities.Realm(
...     realm_proxy=smart_proxy,
...     realm_type='Active Directory',
...     name='foo2',
... ).create_json(create_missing=False)
>>> attrs['id']
2
>>> realm = entities.Realm(id=2).read()
>>> realm.id, realm.name
(2, u'foo2')
```

Comment (Ichimonji10) [inline nailgun/entities.py]: The "robustness" branch I keep talking about provides some nicer error messages, among other things. See:

```python
>>> import robottelo; from nailgun import entities
>>> smart_proxy = entities.SmartProxy(id=1).read()
>>> realm = entities.Realm(
...     realm_proxy=smart_proxy,
...     realm_type='Active Directory',
...     name='foo3',
... ).create(create_missing=False)
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 837, in create
    return self.read(attrs=self.create_json(create_missing))
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 692, in read
    in _get_entity_ids(field_name, attrs)
  File "/home/ichimonji10/code/nailgun/nailgun/entity_mixins.py", line 274, in _get_entity_ids
    attrs.keys()
nailgun.entity_mixins.MissingValueError: Cannot find a value for the "organization" field. Searched for keys named ('organization_ids', 'organization', 'organizations'), but available keys are [u'realm_proxy_id', u'name', u'updated_at', u'created_at', u'hosts_count', u'id', u'hostgroups_count', u'realm_type'].
```

File: nailgun/entities.py (hunk 1 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: I've created #115 to track this.

**Comment by oshtaier** [comment]
Looking forward to see your "robustness" branch in action! :smiley_cat:

**Comment by elyezer** [comment]
ACK. Thanks for the PR and for increasing the coverage. You did a excellent job.

**Pull request merged**